### PR TITLE
core(gather): define new DNS failure LH Error

### PR DESCRIPTION
**Pull request by paulirish**
Seeing a lot of these errors. I think we want to expose them to the user individually.

### CODE REVIEW

File: lighthouse-core/lib/lh-error.js [hunk 1]
```diff
@@ -182,6 +182,13 @@ const ERRORS = {
     lhrRuntimeError: true,
   },
 
+  // Protocol timeout failures
+  DNS_FAILURE: {
```
Comment (patrickhulce) [inline lighthouse-core/lib/lh-error.js]: proto update :)

**Comment by paulirish** [comment]
ptal

File: proto/lighthouse-result.proto [hunk 2]
```diff
@@ -49,6 +49,8 @@ enum LighthouseError {
   PROTOCOL_TIMEOUT = 17;
   // Used when the page is not responding after maxWaitForLoad.
   PAGE_HUNG = 18;
+  // DNS timed out for the main document request.
```
Comment (brendankenny) [inline proto/lighthouse-result.proto]: handles more cases than just a timeout?

File: lighthouse-core/gather/gather-runner.js [hunk 3]
```diff
@@ -150,8 +150,19 @@ class GatherRunner {
...
       errorDef = LHError.errors.NO_DOCUMENT_REQUEST;
     } else if (mainRecord.failed) {
-      errorDef = {...LHError.errors.FAILED_DOCUMENT_REQUEST};
-      errorDef.message += ` ${mainRecord.localizedFailDescription}.`;
+      const netErr = mainRecord.localizedFailDescription;
+      // Match all resolution and DNS failures
+      // https://cs.chromium.org/chromium/src/net/base/net_error_list.h?rcl=cd62979b
+      if (
+        netErr.startsWith('net::ERR_NAME_NOT_RESOLVED') ||
+        netErr.startsWith('net::ERR_NAME_RESOLUTION_FAILED') ||
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: just test equality for these two? Or are these expected as prefixes?

File: lighthouse-core/gather/gather-runner.js [hunk 4]
```diff
@@ -150,8 +150,19 @@ class GatherRunner {
...
+        netErr.startsWith('net::ERR_DNS_')
+      ) {
+        errorDef = {...LHError.errors.DNS_FAILURE};
+      } else {
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: these conditionals are multiplying in an unseemly fashion. Maybe #6457 can take on cleaning them up with early returns @exterkamp

File: lighthouse-core/gather/gather-runner.js [hunk 5]
```diff
@@ -150,8 +150,19 @@ class GatherRunner {
...
-      errorDef.message += ` ${mainRecord.localizedFailDescription}.`;
+      const netErr = mainRecord.localizedFailDescription;
+      // Match all resolution and DNS failures
+      // https://cs.chromium.org/chromium/src/net/base/net_error_list.h?rcl=cd62979b
+      if (
+        netErr.startsWith('net::ERR_NAME_NOT_RESOLVED') ||
+        netErr.startsWith('net::ERR_NAME_RESOLUTION_FAILED') ||
+        netErr.startsWith('net::ERR_DNS_')
+      ) {
+        errorDef = {...LHError.errors.DNS_FAILURE};
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: if no string/message altering, no need to object spread to get a copy (e.g. see `NO_DOCUMENT_REQUEST` above)

**Comment by paulirish** [comment]
@patrickhulce @exterkamp sg?

File: lighthouse-core/gather/gather-runner.js [hunk 6]
```diff
@@ -336,10 +348,10 @@ class GatherRunner {
       passContext.options = gathererDefn.options || {};
 
       // If there was a pageLoadError, fail every afterPass with it rather than bail completely.
-      const artifactPromise = pageLoadError ?
-        Promise.reject(pageLoadError) :
-        // Wrap gatherer response in promise, whether rejected or not.
-        Promise.resolve().then(_ => gatherer.afterPass(passContext, passData));
+      const artifactPromise = pageLoadError
+        ? Promise.reject(pageLoadError)
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: did prettier get loose? This is bad (I'm surprised we don't lint it) and the concat thing above is definitely worse :)

Comment (paulirish) [reply lighthouse-core/gather/gather-runner.js]: prettier! get back in your cage.

**Pull request merged**